text/0006-wasm-smart-contracts.md

```diff
@@ -0,0 +1,440 @@
+---
+layout: default
+title: Wasm_Smart_Contracts
+parent: RFCs
+---
+
+- Feature Name: Support writing Smart Contracts to be executed within a WebAssembly runtime
+- Start Date: 2020-05-20
+- RFC PR: (leave this empty)
+- Fabric Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+This RFC is providing:
+
+* a WebAssembly runtime that can be used as a Chaincode implementation in Fabric
+* implementation of a set of APIs to permit Smart Contracts to be written in Rust and then compiled to Wasm bytecode
+* documented interface to permit the community to implement Smart Contracts in other languages. 
+
+A Rust developer would write their Smart Contract with Rust and produce a library. This they would then target to be compiled to Wasm bytecode - the resulting wasm binary would then be deployed using Fabric v2 "External Builders" into a Wasm runtime. This contract is then treated just the same as any other contract or chaincode. 
+
+The APIs will be implemented at a level consistent with the existing Contract APIs but taking the opportunity to make improvements. 
+
+Note: code examples are provided for illustration, and should be considered pseudo-code to outline the functionality that will be produced.
+
+Prototypes are available for the [Rust API](https://github.com/hyperledgendary/fabric-contract-api-rust), [Wasm Chaincode](https://github.com/hyperledgendary/fabric-chaincode-wasm); work is still taking place in these repos so please get contact with questions.
+
+# Motivation
+[motivation]: #motivation
+
+Wasm provides a "well defined, tightly specified, isolated" runtime. It is ideally suited to executing code in a well-controlled and safe manner. Originally intended for use within the browser it is perfectly suited to running outside the browser. The requirements of chaincode and smart contracts are ideally suited to the Wasm model. 
```

Comment:
> Wasm provides a "well defined, tightly specified, isolated" runtime. It is ideally suited to executing code in a well-controlled and safe manner. 

How do you see the typical use of containerization of the chaincode runtime we have today? Is is contradictory or complementary to WASM's isolation property?

Reply:
Complementary; today the sets of languages suitable for compilation to Wasm have minimal overlap.

Many Wasm runtimes are planning features such as limited execution limits, so could be halted.  

Early days as well for Wasm,.